clis/teliod/src/qnap.rs

```diff
@@ -0,0 +1,398 @@
+}
+
+fn teliod_socket_exists() -> bool {
+    DaemonSocket::get_ipc_socket_path()
+        .and_then(|path| path.metadata())
+        .is_ok()
+}
+
+fn kill_teliod_process() -> Result<(), io::Error> {
+    match fs::read_to_string(PID_FILE) {
+        Ok(pid) => {
+            let _ = Command::new("kill").arg("-9").arg(pid.trim()).status();
+            if teliod_socket_exists() {
+                let _ = DaemonSocket::get_ipc_socket_path().and_then(fs::remove_file);
+            }
+            fs::remove_file(PID_FILE)
+        }
+        Err(e) => Err(e),
+    }
+}
+
+fn start_daemon() -> Response {
+    if is_teliod_running() {
+        if teliod_socket_exists() {
+            return text_response(400, "Application is already running.");
+        } else {
+            let _ = kill_teliod_process();
+        }
+    }
+
+    let teliod_log_file = match fs::OpenOptions::new()
+        .create(true)
+        .write(true)
+        .truncate(true)
+        .open(TELIOD_LOG)
+    {
+        Ok(file) => file,
+        Err(_) => {
+            return text_response(500, "Failed to open teliod log file.");
+        }
+    };
+
+    match Command::new("setsid")
+        .arg(TELIOD_BIN)
+        .arg("daemon")
+        .arg(TELIOD_CFG)
+        .stdout(Stdio::from(
+            teliod_log_file
+                .try_clone()
+                .expect("Failed to open teliod log file."),
+        ))
+        .stderr(Stdio::from(teliod_log_file))
+        .spawn()
+    {
+        Ok(child) => {
+            let pid = child.id();
+            if fs::write(PID_FILE, pid.to_string()).is_ok() {
+                text_response(200, "Application started successfully.")
+            } else {
+                text_response(500, "Failed to save PID.")
+            }
+        }
+        Err(_) => text_response(500, "Failed to start the application."),
+    }
+}
+
+fn stop_daemon() -> Response {
+    if is_teliod_running() {
+        if kill_teliod_process().is_ok() {
+            text_response(200, "Application stopped successfully.")
+        } else {
+            text_response(500, "Failed to stop the application.")
+        }
+    } else {
+        text_response(400, "Application is not running.")
+    }
+}
+
+fn update_config(body: &str) -> Response {
+    let mut config: TeliodDaemonConfig = match fs::read_to_string(TELIOD_CFG)
+        .and_then(|content| serde_json::from_str(&content).map_err(|e| e.into()))
+    {
+        Ok(config) => config,
+        Err(e) => {
+            eprintln!("Error reading config file: {}", e);
+            return text_response(500, "Failed to read existing config");
+        }
+    };
+
+    let update_values: serde_json::Value = match serde_json::from_str(body) {
+        Ok(updates) => updates,
+        Err(e) => {
+            eprintln!("Error parsing config json: {}", e);
+            return text_response(400, "Invalid JSON payload");
+        }
+    };
+
+    if let serde_json::Value::Object(update_map) = update_values {
```

Comment:
Please define a type that represents valid body and use (de)serialisation derived for it instead of manual json value traversal.

Reply:
added